x/ecocredit/server/basket/take.go

```diff
@@ -2,11 +2,185 @@ package basket
 
 import (
 	"context"
+	"fmt"
 
+	"github.com/regen-network/regen-ledger/types"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/regen-network/regen-ledger/types/math"
+
+	basketv1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/basket/v1"
 	baskettypes "github.com/regen-network/regen-ledger/x/ecocredit/basket"
 )
 
-func (k Keeper) Take(ctx context.Context, take *baskettypes.MsgTake) (*baskettypes.MsgTakeResponse, error) {
-	//TODO implement me
-	panic("implement me")
+func (k Keeper) Take(ctx context.Context, msg *baskettypes.MsgTake) (*baskettypes.MsgTakeResponse, error) {
+	basket, err := k.stateStore.BasketStore().GetByBasketDenom(ctx, msg.BasketDenom)
+	if err != nil {
+		return nil, err
+	}
+
+	retire := msg.RetireOnTake
+	if !basket.DisableAutoRetire && !retire {
+		return nil, ErrCantDisableRetire
+	}
+
+	amountBasketTokens, ok := sdk.NewIntFromString(msg.Amount)
+	if !ok {
+		return nil, sdkerrors.ErrInvalidRequest.Wrapf("bad integer %s", msg.Amount)
+	}
+
+	acct, err := sdk.AccAddressFromBech32(msg.Owner)
+	if err != nil {
+		return nil, err
+	}
+
+	sdkContext := sdk.UnwrapSDKContext(ctx)
+	basketCoins := sdk.NewCoins(sdk.NewCoin(basket.BasketDenom, amountBasketTokens))
+	err = k.bankKeeper.SendCoinsFromAccountToModule(sdkContext, acct, baskettypes.BasketSubModuleName, basketCoins)
+	if err != nil {
+		return nil, err
+	}
+
+	err = k.bankKeeper.BurnCoins(sdkContext, baskettypes.BasketSubModuleName, basketCoins)
+	if err != nil {
+		return nil, err
+	}
+
+	amountBasketTokensDec, err := math.NewDecFromString(msg.Amount)
+	if err != nil {
+		return nil, err
+	}
+
+	multiplier := math.NewDecFinite(1, int32(basket.Exponent))
+	amountCreditsNeeded, err := amountBasketTokensDec.QuoExact(multiplier)
+	if err != nil {
+		return nil, err
+	}
+
+	var credits []*baskettypes.BasketCredit
+	for {
+		it, err := k.stateStore.BasketBalanceStore().List(ctx,
+			basketv1.BasketBalanceBasketIdBatchStartDateIndexKey{}.WithBasketId(basket.Id),
+		)
```

Comment:
shouldn't this be before the `for` loop?

Reply:
No we reopen the iterative each time. You cannot update/delete the item the iterative points to while it's open. Trying to allow mutable iterators upstream in the orm